### PR TITLE
Fix SDP when upgrading/downgrading video

### PR DESCRIPTION
**Pull request by manuquentin**

### CODE REVIEW

**Review by bogue — DISMISSED** [review]

**Comment by sboily** [comment]
Please add information in README before merge this PR.

File: src/domain/Phone/WebRTCPhone.js [hunk 1]
```diff
@@ -184,14 +184,36 @@ export default class WebRTCPhone extends Emitter implements Phone {
...
 
-    return this.client.reinvite(sipSession, newConstraints);
+    const callSession = this.callSessions[this.getSipSessionId(sipSession)];
+    const hasRemoteVideo = callSession ? !!this.getRemoteVideoReceiver(callSession) : false;
+    // Do not recreate a stream in 1:1 when a remote video stream is already present. Just replace the video track.
+    if (callSession && constraints && constraints.video && hasRemoteVideo && !conference) {
+      // Reset session video device id to create new stream
+      await this.changeSessionVideoInputDevice(null);
+      // $FlowFixMe
+      await this.client.setVideoInputDevice(typeof newConstraints.video === 'string' ? newConstraints.video
```
Comment (sduthil) [inline src/domain/Phone/WebRTCPhone.js:202]: What kind of string value can `newConstraints.video` take? Shouldn't it be a boolean or an object?

File: src/simple/room/Room.js [hunk 2]
```diff
@@ -346,6 +350,25 @@ class Room extends Emitter {
...
+    logger.info('send room reinvite', { sipSession, newConstraints });
+
+    await Wazo.Phone.phone.sendReinvite(sipSession, newConstraints, true);
+
+    if (this.localParticipant && newConstraints && newConstraints.video) {
+      const pc = sipSession.sessionDescriptionHandler.peerConnection;
+      pc.getSenders()
+        .filter(localSender => localSender.track && localSender.track.kind === 'video')
+        .forEach(localVideoSender => {
+          const mediaStream = new MediaStream();
```
Comment (sduthil) [inline src/simple/room/Room.js:363]: I'm not sure it should be a new `MediaStream`: it should be synchronized with audio sender, so IMHO it should be the same `MediaStream` than audio, but I couldn't find how to get the same `MediaStream`

File: src/utils/sdp.js [hunk 3]
```diff
@@ -37,6 +37,102 @@ export const isSdpValid = (sdp: ?string): boolean => {
...
+  const bundleIndex = parsedSdp.groups.findIndex(group => group.type === 'BUNDLE');
+  if (bundleIndex !== -1) {
+    parsedSdp.groups[bundleIndex].mids = parsedSdp.media
+      .map((media, index) => ('mid' in media ? media.mid : index)).join(' ');
+  }
+
+  return sdpParser.write(parsedSdp);
+};
+
+export const randomMid = (sdp: string): string => {
```
Comment (sduthil) [inline src/utils/sdp.js:51]: `randomMid` is unused, should we remove it?

File: src/utils/sdp.js [hunk 4]
```diff
@@ -37,6 +37,102 @@ export const isSdpValid = (sdp: ?string): boolean => {
...
+    return false;
+  }
+
+  const parsedSdp = sdpParser.parse(sdp);
+
+  return !!parsedSdp.media.find(media =>
+    media.type === 'video' && media.port > 10 && (!media.direction || media.direction !== 'inactive'));
+};
+
+export const fixInactiveVideo = (sdp: string): string => {
```
Comment (sduthil) [inline src/utils/sdp.js:102]: Same thing here: the four functions `fixInactiveVideo`, `isLocalVideoInactive`, `disableLocalVideo` and `removeLocalVideo` are unused, should we remove them?

File: src/web-rtc-client.js [hunk 5]
```diff
@@ -833,70 +834,106 @@ export default class WebRTCClient extends Emitter {
...
     if (id === currentId) {
       return null;
     }
 
     // let's update the local value
-    if (this.video && this.video.deviceId) {
-      // $FlowFixMe
-      this.video.deviceId.exact = id;
+    if (this.video) {
+      this.video = {
```
Comment (sduthil) [inline src/web-rtc-client.js:853]: This may override other video constraints, no?

File: src/web-rtc-client.js [hunk 6]
```diff
@@ -833,70 +834,106 @@ export default class WebRTCClient extends Emitter {
...
     if (newConstraints) {
       this.changeVideo(!!newConstraints.video);
     }
 
+    // When upgrading to video, remove the `deactivateVideoModifier` modifiers
+    if (newConstraints && newConstraints.video) {
+      const modifiers = sipSession.sessionDescriptionHandlerModifiersReInvite;
+      sipSession.sessionDescriptionHandlerModifiersReInvite = modifiers.filter(modifier =>
+        modifier !== deactivateVideoModifier);
+      sipSession.sessionDescriptionHandlerModifiersReInvite.push(activateVideoModifier);
```
Comment (sduthil) [inline src/web-rtc-client.js:929]: Are you sure we need an additional modifier? Isn't it enough to remove the `deactivateVideoModifier`?

File: src/simple/room/Room.js (hunk 2 above)
Comment (manuquentin) [reply src/simple/room/Room.js:363]: We'll take a look when dealing with reinvite in SFU mode.

File: src/web-rtc-client.js (hunk 5 above)
Comment (manuquentin) [reply src/web-rtc-client.js:853]: Yes that's the point. Here we want only to change the video device id (without changing it in the session), so we have to change the constraint.
Changing the constraint in the session is done in `changeVideoInputDevice` .

File: src/domain/Phone/WebRTCPhone.js (hunk 1 above)
Comment (sduthil) [reply src/domain/Phone/WebRTCPhone.js:202]: For the record: The boolean/object format is only for the low-level (WebRTC) constraints. The higher-level Wazo SDK accepts constraints as boolean or string, where the string is the device ID.

**Pull request merged**